### DISCUSSION TITLE
Review of {admiraldev} Developer Guide pages ahead of 1.0

### DISCUSSION THREAD

**Discussion by manciniedoardo** [pharmaverse/admiral#2183] [General]
Please share your thoughts about what should be enhanced/modified/removed ahead of 1.0.

**Comment by manciniedoardo** [comment]
# Development Process page - list your comments below.

**Reply by bms63** [reply]
* Wondering if this page could just be moved into https://pharmaverse.github.io/admiral/cran-release/articles/contribution_model.html as we repeat a lot of the same information. 

*  I am wondering if we really need to list out the checklist for the Pull Request Template.  We could just say check out the latest template checks and link to it.

* We have the handy youtube video for contributing.  It needs to be updated, but we could link it here as well. 

* References to `devel` should be updated.

**Reply by manciniedoardo** [reply]
- Agree with comment above about moving to "Contribute to {admiral}" page. I like that it would be on the main website as well (more accessible to the layman).
- I don't see any harm in retaining the checklist here to be honest. It's a useful reference point.
- Contribution model page referenced here says "Our contribution model is still in testing". I think we can get rid of that now.

**Reply by bms63** [reply]
We will update this

**Reply by bms63** [reply]
Issue Created https://github.com/pharmaverse/admiral/issues/2196

**Comment by manciniedoardo** [comment]
# Programming Strategy page - list your comments below.

**Reply by bms63** [reply]
* Update mandate to strongly encourage 
* Change the unit test to every function should have a unit test
* Not sure what the QC method is here, but I think we could just drop it
* submission ready code is pretty bold - i don;t think this is needed.

![image](https://github.com/pharmaverse/admiral/assets/10111024/a43132bb-3aff-4253-ae0e-c94a84c93efa)

**Reply by bms63** [reply]
Could be nice to have this in a diagram rather than wall of text?  I could mock something up in Lucid.   

![image](https://github.com/pharmaverse/admiral/assets/10111024/c904ca90-7ab6-4e65-90f1-c2bcce8fba9e)

**Reply by bms63** [reply]
Is this still true?  Seems like we allow global objects now??

![image](https://github.com/pharmaverse/admiral/assets/10111024/7f940ae3-22b8-45c1-8bce-682ee8a821f6)

**Reply by bms63** [reply]
Unless specified by the user?

![image](https://github.com/pharmaverse/admiral/assets/10111024/9bdcaca0-d80b-47ff-bf96-e4982a3f46df)

**Reply by bms63** [reply]
Investigate breakage of cross-linking? 
![image](https://github.com/pharmaverse/admiral/assets/10111024/715ba28e-a99c-4585-a1ea-f9bb3f80c844)

**Reply by bms63** [reply]
We should encourage functions with 10 arguments or less - exceptions can be made, but I think we should rethink a function with 10+ arguments.  

![image](https://github.com/pharmaverse/admiral/assets/10111024/dd4e193c-ad37-4d39-a572-471bed75cbdd)

**Reply by bms63** [reply]
For Metadata - we do ingest stuff for Lab Grading now.  I think we should just remove this section.  
![image](https://github.com/pharmaverse/admiral/assets/10111024/7a687fd1-b64d-4129-8c0b-85fc78338941)

**Reply by zdz2101** [reply]
I made a draft PR to make the easier to see in context/ the side-by-side https://github.com/pharmaverse/admiraldev/pull/343/files

Highlights would be: 

- I don't think we actually comment as well as we describe in our programming strategy
- some of our keyword/family feel a little outdated/do we ever want to rearrange the categories?
- Sometimes there's R-in-general type of things not necessarily tied to admiral like the conversion of blanks/NAs from SAS, think we would benefit from removal and pointing people to the fantastic textbooks that achieve this
- file structuring section doesn't feel aligned with what we actually do
- I don't know if we technically do the whole `suppress_warning` portion of step 1 of deprecation as described

**Reply by bms63** [reply]
> I made a draft PR to make the easier to see in context/ the side-by-side https://github.com/pharmaverse/admiraldev/pull/343/files
> 
> Highlights would be:
> 
> * I don't think we actually comment as well as we describe in our programming strategy

Well...more of a goal to reach :) 

> * some of our keyword/family feel a little outdated/do we ever want to rearrange the categories?

Yes happy to re-visit.   I think since we are more "stable" that a revisit makes a lot of sense.  

> * Sometimes there's R-in-general type of things not necessarily tied to admiral like the conversion of blanks/NAs from SAS, think we would benefit from removal and pointing people to the fantastic textbooks that achieve this

Not sure I am following. 

> * file structuring section doesn't feel aligned with what we actually do

Yeah - I think we should re-visit this in a meeting.

> * I don't know if we technically do the whole `suppress_warning` portion of step 1 of deprecation as described

Shall we update to our current practice?

**Reply by manciniedoardo** [reply]
Functional Workflow:
- Admiral should be corrected to `admiral`

Functions in R:
- admiraldev should be corrected to `admiraldev`

Deprecation:
- Rename to deprecation/supercession (is that a word?) and outline our strategy for superseding functions

**Reply by bundfussr** [reply]
> Is this still true? Seems like we allow global objects now??
> 
> ![image](https://user-images.githubusercontent.com/10111024/277677656-7f940ae3-22b8-45c1-8bce-682ee8a821f6.png)

I would consider this still as true. At the moment we use the objects from the `admiral_environment` as default values only. So they are still  inputs to the function.

**Reply by bundfussr** [reply]
> Unless specified by the user?
> 
> ![image](https://user-images.githubusercontent.com/10111024/277676362-9bdcaca0-d80b-47ff-bf96-e4982a3f46df.png)

I see not need to change. I am not aware of admiral functions which return the dataset with a specific sort order.

**Comment by manciniedoardo** [comment]
# Writing Vignettes  page - list your comments below.

**Reply by bms63** [reply]
I don't think this needs to exist or could be consolidated into another Vignette with the Unit Tests

**Reply by bms63** [reply]
Actually --- think we should just remove it and consolidate Unit Tests into Programming Strategy.

**Reply by manciniedoardo** [reply]
I'm playing devil's advocate here but is it really necessary to completely remove this "Writing Vignettes" page? The content is pretty much exactly exemplifying the conventions we use, and even if we don't reference this page that often, what's the harm in keeping it? If it could be useful for one dev every once in a while then it would be a shame to get rid of it.

**Reply by bms63** [reply]
Just trying to cut down on the amount of reading a new dev needs to do.  We don't really have anything admiral-flavored here so was thinking a new dev would just need to read the r-pkg manual to become familiar with vignettes

**Reply by manciniedoardo** [reply]
I would argue that portions such as this one could be quite useful for, say, an extension package team just starting out. Although, granted, you could just look at an existing vignette and copy that...
![image](https://github.com/pharmaverse/admiral/assets/53403957/5932d77d-e0e0-4ef9-aa12-929aa2301ab9)

**Reply by bms63** [reply]
I think it could just go into the Programming Strategy as a section.

**Reply by manciniedoardo** [reply]
Can do, although we will be making the programming strategy quite long!

**Reply by bms63** [reply]
Going to leave as is

**Comment by manciniedoardo** [comment]
# Unit Test Guidance  page - list your comments below.

**Reply by bms63** [reply]
We have some interesting flavors with our Unit Tests, but I don't think it needs its own vignette. Can we move relveant admiral flavors into Programming Strategy? There is already a section for it?

**Reply by manciniedoardo** [reply]
I agree with this. In particular I think the first two sections ("Why Write Unit Tests" and "Writing Good Unit Tests") are not specific to {admiral} so they should be removed. In fact, from the third section I would only retain information about the add-in.

**Reply by bms63** [reply]
Well then we will have room for some more vignettes that I know @ddsjoberg wants to write!!

**Reply by bms63** [reply]
Going to keep and move the Programming Strategy testing stuff into here

**Comment by manciniedoardo** [comment]
# Guidance for git and github usage page - list your comments below.

**Reply by manciniedoardo** [reply]
The "Introduction" and "Branches" and "Working with Feature Branches" are out of date. Now that we are working with the githubflow strategy maybe we can just reference to that in the "Contributing to {admiral}" page on the main website and do away with this page entirely.

**Reply by bundfussr** [reply]
The vignette needs to be updated but I would not remove it. I think the naming conventions for branches and commit messages are important and we need to add something regarding version numbering and update of `DESCRIPTION` (`Remotes:` field and using .9000 in required version).

**Comment by manciniedoardo** [comment]
# PR Review guidance  page - list your comments below.

**Reply by manciniedoardo** [reply]
Generally I think we should retain this page, with the modifications below:

- We should remove references to `devel` here
- Link to "Development Process" page will have to be updated if we do end up moving that page
- The "GitHub Actions/CI Workflows" section can be shortened and we can just link to the [admiralci website](https://pharmaverse.github.io/admiralci/main/) (which should be updated ahead of 1.0)

**Comment by manciniedoardo** [comment]
# R CMD Issues page  - list your comments below.

**Reply by manciniedoardo** [reply]
I think the contents of this page should be updated and moved to the [admiralci website](https://pharmaverse.github.io/admiralci/main/)

**Comment by manciniedoardo** [comment]
# Release strategy page - list your comments below.

**Reply by manciniedoardo** [reply]
I think this page could be superseded by linking to a page exemplifying the githubflow release strategy (if there is one?)

**Reply by zdz2101** [reply]
https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow#:~:text=The%20overall%20flow%20of%20Gitflow%20is%3A%201%201.,is%20merged%20into%20develop%20and%20main%20More%20items

Atlassian is certainly a leader in the space, going to expect a lot of organizations are using their enterprise software suite so probably used to reading their documentation too

**Reply by bundfussr** [reply]
I think the steps for normal releases and hotfix releases are quite useful and it is unlikely that we find an external page with this information. An update is required due to the change of the branching strategy (see https://github.com/pharmaverse/admiraldev/pull/330 for a draft).

**Comment by manciniedoardo** [comment]
# Test Data Guidance  page - list your comments below.

**Reply by manciniedoardo** [reply]
This page was recently updated, I see no reason for changes here :)

**Comment by manciniedoardo** [comment]
# Package Extensions page - list your comments below.

**Reply by manciniedoardo** [reply]
This page was recently updated, I see no reason for changes here either

**Comment by bms63** [comment]
# Admiral Manifesto Revisited

**Reply by bms63** [reply]
I think this should be made into a infographic on admiral and the text should live on the readme in admiraldev.  The infographic should link to the admiraldev readme.

**Reply by manciniedoardo** [reply]
Agree with this suggestion.

**Comment by rossfarrugia** [comment]
Good discussions team! Adding some high level thoughts in case useful here @bms63 and @manciniedoardo - overall i think a lot of work could potentially be created here updating these pages but i would focus on value updates here (e.g. where our dev process has changed like git branching strategy) rather than an overhaul given that i expect we shouldn't need as many new community contributors compared to when we were first building up this package. also there was a few mention of removing some pages that gave more generic guidance such as unit testing - i get that as they're not related specifically to admiral and maybe one day we'd move these to some higher level pharmaverse location, but fyi i have been sharing these with other pharmaverse pkg dev teams like datacutr and sdtmchecks. people have found them really useful, so i'd be in favour of not removing any content for now and down the line if we find a better place to store it then we could move. p.s. tag me in any reply if you'd like my further input, as i won't ask for notifications on this thread for fear of an already busy inbox exploding 😄

**Reply by bms63** [reply]
Okay sounds good on keeping the vignettes.  I was just trying to cut down on paperwork that is already repeated in the r-pkg manual.  

Understood on not trying to do a serious overhaul.  I am mostly interested in removing then re-writing and making sure there is nothing contradictory in our strategy.

**Comment by bms63** [comment]
# Revamp of admiraltemplate

**Reply by bundfussr** [reply]
Do we need admiraltemplate? Do we expect more admiral extension packages?

**Reply by bms63** [reply]
I think it is worthwhile to keep in action.   I know a few other non-admiral projects made use of the template.  It will need a bit of touch up  post-1.0 and renv removal.

**Reply by bms63** [reply]
but for other extensions - that is the dream, but I don't know of any in the queue.

**Reply by manciniedoardo** [reply]
Agree we should keep - we have no extension packages currently in the works but this could change at any moment if the nee /opportunity arises or an interested individual/company comes forward.
